### PR TITLE
[Console] Use readline for user input when available #DX

### PR DESCRIPTION
**Pull request by michaelperrin**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

Given I am entering data in an user input
When I use left and right keys to make some changes in what I have typed
Then the cursor should move accordingly instead of adding characters at the end of the line

To make it simple: using the arrow keys (←  →)  to make changes to what I already typed would be much handier than getting `^[[D` and `^[[C` characters in the terminal and having to delete all chars to type everything again.

I could not add any extra tests to this as the STDIN can't be used during tests. But they are not breaking and I tried again all types of questions (text, choices, hidden) by myself.

Note that `readline` can't be used for hidden questions, as `stty -echo` is not taken into account.

### CODE REVIEW

File: src/Symfony/Component/Console/Helper/QuestionHelper.php [hunk 1]
```diff
@@ -438,4 +460,14 @@ private function hasSttyAvailable()
...
     }
+
+    /**
+     * Returns whether the GNU Readline library is available or not.
+     *
+     * @return bool
+     */
+    private function hasReadline()
+    {
+        return function_exists('readline');
```
Comment (fabpot) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: I would inline that instead of creating a method for that.

**Comment by fabpot** [comment]
:+1:

**Comment by michaelperrin** [comment]
@fabpot I inlined the check on the availability of readline.

File: src/Symfony/Component/Console/Helper/QuestionHelper.php [hunk 2]
```diff
@@ -424,6 +420,32 @@ private function getShell()
...
+        if (STDIN === $stream && function_exists('readline')) {
+            $ret = readline();
+        } else {
+            $ret = fgets($stream, 4096);
+
+            if (false === $ret) {
+                throw new \RuntimeException('Aborted');
+            }
+
+            $ret = trim($ret);
```
Comment (xabbuh) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: What about using `stream_get_line()` for platforms not supporting `readline()`? We could then get rid of the `trim()` call.

Comment (xabbuh) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: By the way, the behaviour is now inconsistent as the string returned by `readline()` is not trimmed (for example, when containing extra spaces before or after the actual input).

**Comment by xabbuh** [comment]
Status: Needs work

File: src/Symfony/Component/Console/Helper/QuestionHelper.php (hunk 2 above)
Comment (michaelperrin) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php]: Thanks for your review.
I didn't know about `stream_get_line()` method. Would there be an advantage of using it on platforms not supporting `readline()`? The string would need to be trimmed as well if we want to keep the same behavior (removing whitespaces around the string) and arrow keys can't be used as well, so it doesn't seem there is any benefit.

I agree with you on the second point I should trim the string both with `readline` and with the usual `fgets` functions.

**Comment by michaelperrin** [comment]
@xabbuh I made the change according to you second comment. My comments about your first one are above.

**Comment by xabbuh** [comment]
@michaelperrin I could imagine that people find `stream_get_line()` more readable. But I don't have any hard feelings about it. :+1: 

Status: Reviewed

**Comment by fabpot** [comment]
Thank you @michaelperrin.

**Comment by michaelperrin** [comment]
Thank you!

**Comment by stof** [comment]
this is missing the handling of ``false`` return value when readline is used

**Comment by michaelperrin** [comment]
@stof `readline()` never returns `false` (does it?), that's why I didn't handled this value.

**Comment by stof** [comment]
The comments on the documentation page of the function are saying it does: http://php.net/manual/en/function.readline.php#86719

**Comment by michaelperrin** [comment]
@stof You're right! Too bad it was hidden in the PHP doc comments. I opened #15450

**Pull request merged**